fdbclient/include/fdbclient/MetaclusterManagement.actor.h

```diff
@@ -1062,6 +1055,315 @@ Future<Void> managementClusterRemoveTenantFromGroup(Transaction tr,
+		}
+
+		return Void();
+	}
+
+	ACTOR static Future<Optional<TenantName>> lookupTenantInManagementClusterById(RestoreClusterImpl* self) {
+		Optional<TenantName> tenantName =
+		    wait(self->ctx.runManagementTransaction([self = self](Reference<typename DB::TransactionT> tr) {
+			    return ManagementClusterMetadata::tenantMetadata().tenantIdIndex.get(tr, self->tenantId);
+		    }));
+
+		return tenantName;
+	}
+
+	ACTOR static Future<Optional<TenantName>> lookupTenantInDataClusterById(RestoreClusterImpl* self) {
+		Optional<TenantName> tenantName =
+		    wait(self->ctx.runDataClusterTransaction([self = self](Reference<ITransaction> tr) {
+			    return TenantMetadata::tenantIdIndex().get(tr, self->tenantId);
+		    }));
+
+		return tenantName;
+	}
+
+	ACTOR static Future<Void> renameTenantOnDataCluster(RestoreClusterImpl* self) {
+		ASSERT(self->tenantId != -1);
+		wait(self->ctx.runDataClusterTransaction([self = self](Reference<ITransaction> tr) {
+			return TenantAPI::renameTenantTransaction(
+			    tr, self->tenantName, self->tenantNewName, self->tenantId, ClusterType::METACLUSTER_DATA);
+		}));
+		return Void();
+	}
+
+	ACTOR static Future<Void> removeTenantFromDataCluster(RestoreClusterImpl* self) {
+		wait(self->ctx.runDataClusterTransaction([self = self](Reference<ITransaction> tr) {
+			return TenantAPI::deleteTenantTransaction(
+			    tr, self->tenantName, self->tenantId, ClusterType::METACLUSTER_DATA);
+		}));
+		return Void();
+	}
+
+	// This only supports the restore of an already registered data cluster, for now.
+	ACTOR static Future<Void> run(RestoreClusterImpl* self) {
+		state bool clusterIsRestoring;
+		// set state to restoring
+		try {
+			wait(store(clusterIsRestoring,
+			           self->ctx.runManagementTransaction([self = self](Reference<typename DB::TransactionT> tr) {
+				           return markClusterAsRestoring(self, tr);
+			           })));
+		} catch (Error& e) {
+			// If the transaction retries after success or if we are trying a second time to restore the cluster, it
+			// will throw an error indicating that the restore has already started
+			if (e.code() == error_code_cluster_restored) {
+				clusterIsRestoring = true;
+			} else {
+				throw;
+			}
+		}
+
+		if (!clusterIsRestoring) {
+			// get all the tenant information from the newly registered data cluster
+			wait(getAllTenantsFromDataCluster(self));
+			// get all the tenant information for this data cluster from manangement cluster
+			wait(getAllTenantsFromManagementCluster(self));
+
+			state std::unordered_map<TenantName, TenantMapEntry>::iterator itr = self->dataClusterTenantMap.begin();
+			while (itr != self->dataClusterTenantMap.end()) {
+
+				state TenantName tenantNameOnDataCluster = itr->first;
+				state uint64_t tenantId = (itr->second).id;
+				self->tenantId = tenantId;
+				self->tenantName = tenantNameOnDataCluster;
+
+				state Optional<TenantName> tenantNameOnMgmtCluster = wait(lookupTenantInManagementClusterById(self));
```

Comment:
This is going to be expensive to run all of these checks serially. You could run them in batches so that you could parallelize things a bit, but we could also restructure the logic here accounting for the particular restore scenario being implemented (an older version of a data cluster that was previously on the cluster).

One way to approach this is to:

1. Read the entire tenant map from the data cluster
2. Read the cluster tenant index to get all of the tenants for this data cluster (we probably don't need all tenants everywhere)
3. Read the tenant map entries for all entries in the management tenant index
4. Compare the two lists. Anytime there is a match, we don't have to do anything special.
5. If there is a mismatch, we can resolve that in various ways depending on the type of mismatch. For example, if both entries are present but different, we can apply the changes from the management cluster (and possibly change the management cluster to READY state). If an entry is missing on the management cluster, then we could do a rename check. If still not found, we could optionally do an ID check across the entire metacluster to see if it has been moved, but whether or not it has we would still delete the tenant from this cluster.